internal/stanza/converter.go

```diff
@@ -15,49 +15,294 @@
+	pLogsChan chan pdata.Logs
+
+	stopOnce sync.Once
+	stopChan chan struct{}
+
+	// flushInterval defines how often we flush the aggregated log entries.
+	flushInterval time.Duration
+	// maxFlushCount defines what's the amount of entries in the buffer that
+	// will trigger a flush of log entries.
+	maxFlushCount uint
+	// flushChan is an internal channel used for transporting batched pdata.Logs.
+	flushChan chan []pdata.Logs
+
+	// data is the internal cache which is flushed regularly, either when
+	// flushInterval ticker ticks or when max number of entries for a
+	// particular Resource is reached.
+	data      map[string][]*entry.Entry
+	dataMutex sync.RWMutex
+	dataCount uint
+
+	wg sync.WaitGroup
+
+	logger *zap.Logger
+}
+
+type ConverterOption interface {
+	apply(*Converter)
+}
+
+type optionFunc func(*Converter)
+
+func (f optionFunc) apply(c *Converter) {
+	f(c)
+}
+
+func WithFlushInterval(interval time.Duration) ConverterOption {
+	return optionFunc(func(c *Converter) {
+		c.flushInterval = interval
+	})
+}
+
+func WithMaxFlushCount(count uint) ConverterOption {
+	return optionFunc(func(c *Converter) {
+		c.maxFlushCount = count
+	})
+}
+
+func WithLogger(logger *zap.Logger) ConverterOption {
+	return optionFunc(func(c *Converter) {
+		c.logger = logger
+	})
+}
+
+func NewConverter(opts ...ConverterOption) *Converter {
+	c := &Converter{
+		pLogsChan:     make(chan pdata.Logs),
+		stopChan:      make(chan struct{}),
+		logger:        zap.NewNop(),
+		flushChan:     make(chan []pdata.Logs),
+		flushInterval: DefaultFlushInterval,
+		maxFlushCount: DefaultMaxFlushCount,
+		data:          make(map[string][]*entry.Entry),
+		wg:            sync.WaitGroup{},
+	}
+
+	for _, opt := range opts {
+		opt.apply(c)
+	}
+
+	return c
+}
+
+func (c *Converter) Start() {
+	c.logger.Debug("Starting log converter")
+
+	c.wg.Add(1)
+	go func(c *Converter) {
+		defer c.wg.Done()
+
+		for {
+			select {
+			case <-c.stopChan:
+				return
+
+			case pLogs := <-c.flushChan:
+				if err := c.flush(context.Background(), pLogs); err != nil {
+					c.logger.Error("Problem sending log entries",
+						zap.Error(err),
+					)
+				}
+				// NOTE:
+				// Since we've received a flush signal independently of flush
+				// ticker do we want to reset the flush ticker?
+			}
+		}
+	}(c)
+
+	c.wg.Add(1)
+	go func(c *Converter) {
+		// This goroutine serves ticks of main ticker and signals batch
+		// and conversion of log entries to happen.
+		defer c.wg.Done()
+
+		ticker := time.NewTicker(c.flushInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-c.stopChan:
+				return
+
+			case <-ticker.C:
+				c.dataMutex.RLock()
+				count := c.dataCount
+				c.dataMutex.RUnlock()
+				if count > 0 {
+					pLogs := c.convertBuffer()
+					go c.queueForFlush(pLogs)
+				}
+			}
+		}
+	}(c)
+}
+
+func (c *Converter) Stop() {
+	c.stopOnce.Do(func() {
+		close(c.stopChan)
+		c.wg.Wait()
+		close(c.pLogsChan)
```

Comment:
In the current implementation `pLogsChan` is unbuffered so there won't be any logs to flush from it.

My reasoning here was as follows:

* we close the `stopChan` which signals with internal goroutines to return
* we wait for them to return (to prevent producing dangling goroutines)
* we close the `pLogsChan` to signal the consumers that they can stop consuming (gracefully stopping downstream consumers/pipelines)

If I missed anything or there are reasons to introduce buffering to the channel please let me know.